### ISSUE TITLE
Using Alembic for Single Database with Multiple Schemas

### ISSUE THREAD

**Issue by mohamednajiullah** [sqlalchemy/alembic#710]
**Describe your question**
I'm trying to use alembic to manage database migrations for a postgres database that contains multiple schemas with each schema being used by a different micro-service. 

I realised that I have to use a different table for each micro-service otherwise one micro-service's migration starts deleting other micro-service tables. So I configured each micro-service `alembic_version` table to be in it's own schema to avoid clashes.

The problem that I'm facing is that every migration now recreates all the existing tables of the schema. I tried to debug it and one of the solutions was to set `include_schemas=True`. But this makes alembic scan all other schemas and try to delete all the tables present in them

I'm not sure if I explained the issue well. I can more details as required

**Useful links**
- The [documentation](https://alembic.sqlalchemy.org/en/latest/) website
- The [Cookbook](https://alembic.sqlalchemy.org/en/latest/cookbook.html) wiki
- The SQLAlchemy [documentation](https://docs.sqlalchemy.org/en/latest/) website
- [Stack Overflow](https://stackoverflow.com/questions/tagged/alembic) tag
- Alembic [Google group](https://groups.google.com/forum/#!forum/sqlalchemy-alembic)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment]
you have to decide up front how these schemas are versioned.

in your application, do you have explicit table models written out that each point to just one of the schemas?   this would use a single alembic_version table and include schemas.

or, is each schema a copy of one "master" schema, and your application has just one model that is copied to all of them?  this would be multitenancy, and youd' need to alter your env.py to loop through all the schemas individually.  there would be an alembic_version table in each.  this is a little tricky but I can show you how.

or, you have multiple applciations, each application has table models for just one, or some, of those schemas?  in this case you would also have an alembic_version table in each schema and each application would only look at once schema at a time, if some of those apps use more than one schema you'd need an [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) hook to use the right schemas.

**Comment by mohamednajiullah** [comment]
I can give you an example.

I have a user service that has user details like first name, last name etc. I intend to store the data for the user service in a `user_schema` with a table named `user_details`.

I also have an authentication service that has login details like login_id etc. I intend to store the data for the authentication service in a `auth_schema` with a table named `login_details`.

Each table has its own structure. Therefore I have tried to keep the `alembic_version` table for each service in it's own schema. 

While running migration, alembic looks at that schema alone which is good. However it also tries to recreate the table for every migration even if the table already exists. How do I stop alembic from recreating the existing and migrate only the changes like adding a new column etc..

**Comment by zzzeek** [comment]
> While running migration, alembic looks at that schema alone which is good. However it also tries to recreate the table for every migration even if the table already exists. How do I stop alembic from recreating the existing and migrate only the changes like adding a new column etc..

alembic doesn't do that if it's set up correctly, so we are still not there yet.

so you have:

1. one application 

2. many schemas with different tables

3. you'd like alembic_version in each schema

OK so how would you like to run alembic for each of these separate schemas?   are you running alembic with a single alembic.ini but using the "--name" option ?     How is your env.py set up?   how do you connect to the database, and then do you have an explicit "schema" name in each of your mapped class / Table objects ?

Basically to do what you're doing, a few things need to be done correctly up front and I would need to see how you're doing it.    if you want to run alembic in one remote schema at a time, and that is not the default schema of the database connection,you would definitely need an [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) hook, so that each time alembic runs, it knows which schema to look at.   

this would probably be easier to configure if you ran alembic exactly once for the whole application and for all schemas, you would set include_schemas=True and assuming you have table metadata for every table it would do the right thing.

**Comment by mohamednajiullah** [comment]
Hello @zzzeek ,

I cannot run alembic once for my entire application as the application consists of many independently deployable services. There is a database migration step in the deployment setup which makes the necessary db changes as part of a new deployment.

Each service has it's own `migrations` folder with separate `alembic.ini`s. As for the `env.py`, I've not changed the default file much except add the `version_table_schema` in the following block

```
    with connectable.connect() as connection:
        context.configure(
            connection=connection,
            target_metadata=target_metadata,
            process_revision_directives=process_revision_directives,
            version_table_schema='user_schema',
            **current_app.extensions['migrate'].configure_args
        )

        with context.begin_transaction():
            context.run_migrations()
```

I went through the `include_object` documentation before. I'm struggling to understand how to set it up so that it does not recreate the tables for every migration script.

Thank you for your help!

**Comment by zzzeek** [comment]
OK so you have entirely separate alembic environments each with its own env.py correct?

OK last question do your metadata models have a "schema" in them?  that is, they look like:

```
class MyModel(Base):
   # ...

   __table_args__ = {"schema": "user_schema"}

```      

the include_object we are looking for here looks like this:

```

def include_object(object, name, type_, reflected, compare_to):
    if type_ == "table" and object.schema != "user_schema":
        return False
   else:
        return True

context.configure(
    # ...
    include_schemas=True,
    include_object = include_object
)
```

within the autogenerate process (I'm assuming that's what we're talking about when you say "recreate the tables for every migration script."), you have complete control over what objects are considered by using that hook.

**Comment by mohamednajiullah** [comment]
`OK last question do your metadata models have a "schema" in them? `

Yes @zzzeek . Let me try this out. Thank you so much!!

**Comment by mohamednajiullah** [comment]
It works. 

Thank you so much @zzzeek!!

**Comment by alikazai** [comment]
thanks @zzzeek that fixed my issue

**Comment by lowercase00** [comment]
Hi @zzzeek would you mind sharing some example or pointing to the right direction on setting multi tenancy? I basically have a ser of tables, and each user (tenant) will have the same set (always the same) of tables. So i’m trying to figure a way to control migrations so that a change in one of the SQLAlchemy models will reflect in all users (schemas). Any references that could help? Thanks!

**Comment by zzzeek** [comment]
try that recipe that you just linked on the site, I tested that specifically as something that works for the "postgresql with schema-per-user" design that is fairly common.      you'll see the note included that alembic hasn't had a good story yet for the schema_translate_map feature so the recipe basically skips it.

**Comment by abhianair** [comment]
Hi @zzzeek
I have the same requirement 

>  or, is each schema a copy of one "master" schema, and your application has just one model that is copied to all of them? this would be multitenancy, and youd' need to alter your env.py to loop through all the schemas individually.

Can you explain to me how can I achieve this. Do you have some snippets on this.?

**Comment by zzzeek** [comment]
> Hi @zzzeek
> I have the same requirement
> 
> > or, is each schema a copy of one "master" schema, and your application has just one model that is copied to all of them? this would be multitenancy, and youd' need to alter your env.py to loop through all the schemas individually.
> 
> Can you explain to me how can I achieve this. Do you have some snippets on this.?

it's linked in the comment above, that link is:

https://alembic.sqlalchemy.org/en/latest/cookbook.html#rudimental-schema-level-multi-tenancy-for-postgresql-databases

**Comment by bogdancondurache** [comment]
Hi @zzzeek ,

Thank you for your answers on this topic. I'm trying to achieve something similar to multi-tenancy but for SQL Server and I'm using your SO post linked above. My only issue is that the `alembic_version` is not generated for other schemas so all migrations fail. How can I make sure the table gets created for those schemas as well? I'd preferably like to avoid creating them manually.

**Comment by zzzeek** [comment]
> Hi @zzzeek ,
> 
> Thank you for your answers on this topic. I'm trying to achieve something similar to multi-tenancy but for SQL Server and I'm using your SO post linked above. My only issue is that the `alembic_version` is not generated for other schemas so all migrations fail. How can I make sure the table gets created for those schemas as well? I'd preferably like to avoid creating them manually.

alembic does that for the target database given automatically, so if it's not doing that, then your env.py approach isn't working.

the above recipe was only tested on Postgresql.  SQL Server to my knowledge has no "search path" so unless you're able to make it do that, you'd have to modify the recipe.   Maybe on the SQL server connection you can do something like ["use database"](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/use-transact-sql?view=sql-server-ver15)  - the idea is that alembic runs everything with no schema name known at all, it's all implicit on the given database connection.

> 
> LE: Figured it out. For those that need it, in `context.configure` you have to also add `version_table_schema=schema_name`. 

that's fine but that won't handle all the other tables, so you want to do something like "use schema_name" on the connection each time.

> Now my problem is that migration are still running against the original schema and one of them fails with `Column names in each table must be unique.`. My understanding was that this:
> 
> ```python
>     with connectable.connect() as connection:
>         for schema in supported_schemas:
>             conn = connection.execution_options(schema_translate_map={None: schema})
> 
>             logger.info(f"Migrating schema {schema}")
>             context.configure(
>                 connection=conn,
>                 target_metadata=target_metadata,
>                 process_revision_directives=process_revision_directives,
>                 version_table_schema=schema,
>                 **current_app.extensions["migrate"].configure_args,
>             )
> 
>             with context.begin_transaction():
>                 context.run_migrations()
> ```
> 
> from that SO answer was handling this and making the migration to run in that schema.

Alembic does not use the "schema_translate_map" feature, this is not implemented.   so you need to make it so that a command on the connection automatically applies to the schema you care about each time.

**Comment by bogdancondurache** [comment]
> Alembic does not use the "schema_translate_map" feature, this is not implemented. so you need to make it so that a command on the connection automatically applies to the schema you care about each time.

Ah, got it. My assumption was that it does. Thanks a lot for this! Considering that from what I know SQL Server has no way of doing this easy (with just a query executed), there's no way to achieve it using anything else from alembic, right?

**Comment by zzzeek** [comment]
the multitenancy you're doing is across different SQL Server databases, or schemas?  if databases i would think "USE <databasename>" might work. but! 

another way to do it is to write all your alembic ops using an explicit schema name that you get from a global variable set in your env.py, like:

```
# env.py

import myapp

def run_migrations_online():

    current_tenant = context.get_x_argument(as_dictionary=True).get("tenant")
    with connectable.connect() as connection:
        myapp.current_schema_name = current_tenant

        context.configure(
            connection=connection,
            target_metadata=target_metadata,
        )

        with context.begin_transaction():
            context.run_migrations()
```

then in a migration:

```

import myapp

def upgrade():
    op.add_column('account', sa.Column('last_transaction_date', sa.DateTime), schema=myapp.current_schema_name)
```

**Comment by bogdancondurache** [comment]
> the multitenancy you're doing is across different SQL Server databases, or schemas? if databases i would think "USE " might work. but!

It's in multiple schemas so USE won't help. Most probably I will go for the idea from your snippet, thanks for that!

**Comment by abhianair** [comment]
I did like this. https://abhishek-52727.medium.com/multi-tenancy-in-flask-application-cfdb382b3be

**Comment by bogdancondurache** [comment]
@abhianair Unfortunately that only works for postgres so wasn't helping in my case. But thanks for the link!

@zzzeek Thanks a lot. Managed to solve it with the code above changed to:
```python
        with context.begin_transaction():
            context.get_bind().execute(
                "ALTER USER %s WITH DEFAULT_SCHEMA = %s;" % (connectable.url.username, SCHEMA_NAME)
            )
            context.run_migrations()
            context.get_bind().execute("ALTER USER %s WITH DEFAULT_SCHEMA = %s;" % (connectable.url.username, "dbo"))
 ```
 
Pretty much temporarily changing the default_schema to the one I need and then switching back to dbo. Not the cleanest way 
to do this, but the only one that seems to be working for Sql Server. Hopefully this will help someone else trying to achieve the same thing.

**Comment by henriqueortiz** [comment]
@zzzeek thank you very much, I was trying hard to do the same thing as @mohamednajiullah.

I managed to get Alembic to stop looking at tables from other schemas, however, I still receive some changes to other objects such as sequences.

Is there any way to ignore them too?

It is not generating any migration, but it goes through all this data taking a long time.

![image](https://github.com/sqlalchemy/alembic/assets/36969427/5d234333-b3bf-4d14-a361-93776da447a2)

**Comment by zzzeek** [comment]
That's just a log info that it is finding a sequence that is linked to a column, it's only informational.

you can ignore the whole table, but not that specific hook, unless you use a column reflect event yourself and delete the "default" ahead of time:

```py
import re
from sqlalchemy import event, Table

# global event for all Table objects

@event.listens_for(Table, "column_reflect", insert=True)
def _cancel_default(inspector, table, column_info):
     seq_match = re.match(
                r"nextval\('(.+?)'::regclass\)", column_info["default"]
            )
     if seq_match:
         column_info['default'] = None

```

**Comment by zzzeek** [comment]
adding that yes, after it makes that log, it does a whole query, which can get expensive, for sure.  that hook above should allow it to skip.

**Comment by PaleNeutron** [comment]
Facing the same problem. I create models without schema at local but alembic get my default_schema from connection.

When compare them, alembic think they are different.

Here is the log

```
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected removed foreign key (portfolio_id)(id) on table investorportfolio
INFO  [alembic.autogenerate.compare] Detected removed foreign key (investor_id)(id) on table investorportfolio
INFO  [alembic.autogenerate.compare] Detected added foreign key (portfolio_id)(id) on table investorportfolio
INFO  [alembic.autogenerate.compare] Detected added foreign key (investor_id)(id) on table investorportfolio
INFO  [alembic.autogenerate.compare] Detected removed foreign key (owner_id)(id) on table item
INFO  [alembic.autogenerate.compare] Detected added foreign key (owner_id)(id) on table item
INFO  [alembic.autogenerate.compare] Detected removed foreign key (strategy_id)(id) on table portfolio
INFO  [alembic.autogenerate.compare] Detected added foreign key (strategy_id)(id) on table portfolio
INFO  [alembic.autogenerate.compare] Detected removed foreign key (update_id)(id) on table strategy
INFO  [alembic.autogenerate.compare] Detected removed foreign key (create_id)(id) on table strategy
INFO  [alembic.autogenerate.compare] Detected added foreign key (update_id)(id) on table strategy
INFO  [alembic.autogenerate.compare] Detected added foreign key (create_id)(id) on table strategy
INFO  [alembic.autogenerate.compare] Detected removed foreign key (create_id)(id) on table strategyhis
INFO  [alembic.autogenerate.compare] Detected added foreign key (create_id)(id) on table strategyhis
```
